File: server/common/protocol.py
import socket
from .utils import Bet


OPCODE_INIT_CLIENT = 0x01
OPCODE_BATCH_BETS = 0x02
OPCODE_BATCH_CONFIRMATION = 0x03
OPCODE_FINISHED_BETS = 0x04
OPCODE_REQUEST_WINNERS = 0x05
OPCODE_WAIT = 0x06
OPCODE_WINNERS = 0x07

class Protocol:
    def __init__(self, socket: socket.socket):
        self.socket = socket

    def read_uint8(self) -> int:
        return int.from_bytes(self.read_exact(1), "big")


    def read_uint16(self) -> int:
        return int.from_bytes(self.read_exact(2), "big")


    def write_uint16(self, value: int) -> bytes:
        return value.to_bytes(2, "big")


    def read_exact(self, n: int) -> bytes:
        """Reads exactly n bytes from the socket"""
        buf = b''
        while len(buf) < n:
            chunk = self.socket.recv(n - len(buf))
            if not chunk:
                raise ConnectionError("Connection closed")
            buf += chunk
        return buf


    def read_opcode(self) -> int:
        """
        Reads the opcode from the client socket.
        """
        buf = self.read_exact(1)
        return buf[0]

    def read_string(self) -> str:
        # Read 1 byte of length
        length = self.read_uint8()
        # Read the string bytes
        string_bytes = self.read_exact(length)
        return string_bytes.decode('utf-8')

    def read_bet_msg(self) -> Bet:
        """
        Reads a bet message from the socket.
        Returns a Bet object.
        """

        client_id = self.read_string()
        nombre = self.read_string()
        apellido = self.read_string()
        documento = self.read_string()
        nacimiento = self.read_string()
        numero = self.read_string()

        return Bet(
            agency= client_id, 
            first_name= nombre,
            last_name= apellido,
            document= documento,
            birthdate= nacimiento,
            number= numero )


    def read_batch_msg(self) -> list[Bet]:
        length = self.read_uint16()

        bets = []
        for _ in range(length):
            bet = self.read_bet_msg()
            bets.append(bet)
        return bets


    def send_batch_confirmation(self, success: bool) -> None:
        """
        Sends the batch confirmation to the client without a message or length.
        Format:
            1 byte: opcode = 0x03
            1 byte: success (0 or 1)
        """
        buf = bytearray()
        buf.append(OPCODE_BATCH_CONFIRMATION)
        buf.append(1 if success else 0)
        self.socket.sendall(buf)


    def receive_client_id(self) -> int:
        """
        Receives a message with one byte indicating the length of the client ID, followed by the client ID itself.
        """
        client_id_str = self.read_string()
        client_id = int(client_id_str)
        return client_id


    def send_wait(self) -> None:
        """
        Sends a message to the client indicating that not all agencies have finished yet.
        Format:
            1 byte: opcode (0x06)
        """
        buf = bytearray()
        buf.append(OPCODE_WAIT)
        self.socket.sendall(buf)


    def send_winners(self, winners: list[str]) -> None:
        """
        Sends the winners to the client.
        Format:
            1 byte: opcode (0x07)
            2 bytes: number of winners (uint16 big-endian)
            For each winner:
                1 byte: length of document (uint8)
                N bytes: document (utf-8)
        """
        buf = bytearray()
        buf.append(OPCODE_WINNERS)
        buf.extend(self.write_uint16(len(winners)))

        for document in winners:
            encoded_doc = document.encode('utf-8')
            buf.append(len(encoded_doc))
            buf.extend(encoded_doc)
        self.socket.sendall(buf)

File: client/common/protocol.go
package common

import (
	"encoding/binary"
	"fmt"
	"io"
	"net"
)

const (
	stringLenOverhead = 1 // Overhead for string length prefix

	OpcodeInitClient     = 0x01
	OpcodeBatchBets      = 0x02
	OpcodeFinishedBets   = 0x04
	OpcodeRequestWinners = 0x05
	OpcodeWait           = 0x06
	OpcodeWinners        = 0x07
)

type Protocol struct {
	conn net.Conn
}

func (p *Protocol) sendOpcode(opcode uint8) error {
	return binary.Write(p.conn, binary.BigEndian, opcode)
}

func (p *Protocol) sendString(s string) error {
	if err := binary.Write(p.conn, binary.BigEndian, uint8(len(s))); err != nil {
		return err
	}

	// To avoid short-write and ensure the entire message is sent
	total := 0
	data := []byte(s)
	for total < len(data) {
		n, err := p.conn.Write(data[total:])
		if err != nil {
			return err
		}
		total += n
	}
	return nil
}

func (p *Protocol) sendBet(c *ClientBet, clientId string) error {
	fields := []string{
		clientId,
		c.Nombre,
		c.Apellido,
		c.Documento,
		c.Nacimiento,
		c.Numero,
	}

	for _, field := range fields {
		if err := p.sendString(field); err != nil {
			return err
		}
	}

	return nil
}

func (p *Protocol) sendBatch(batch []ClientBet, clientID string) error {
	// First opcode, then send a message with the batch length, then start sending the bets
	if err := p.sendOpcode(OpcodeBatchBets); err != nil {
		return err
	}

	if err := binary.Write(p.conn, binary.BigEndian, uint16(len(batch))); err != nil {
		return err
	}

	for _, bet := range batch {
		if err := p.sendBet(&bet, clientID); err != nil {
			return err
		}
	}
	return nil
}

// receiveBatchConfirmation receives the batch confirmation from the server.
// Format:
//
//	1 byte opcode (0x03)
//	1 byte: success (0 = fail, 1 = success)

func (p *Protocol) receiveBatchConfirmation() (bool, error) {
	opcode := make([]byte, 1)
	if _, err := io.ReadFull(p.conn, opcode); err != nil {
		return false, err
	}
	if opcode[0] != 0x03 {
		return false, fmt.Errorf("unexpected opcode: %v", opcode[0])
	}

	success := make([]byte, 1)
	if _, err := io.ReadFull(p.conn, success); err != nil {
		return false, err
	}

	return success[0] == 1, nil
}

func (p *Protocol) sizeBytes(clientBet *ClientBet) int {
	fields := []string{
		clientBet.Nombre,
		clientBet.Apellido,
		clientBet.Documento,
		clientBet.Nacimiento,
		clientBet.Numero,
	}

	size := 0
	for _, f := range fields {
		size += stringLenOverhead + len([]byte(f))
	}
	return size
}

func (p *Protocol) sendClientID(clientID string) error {
	if err := p.sendOpcode(OpcodeInitClient); err != nil {
		return err
	}

	return p.sendString(clientID)
}

func (p *Protocol) sendBetsSubmissionCompleted() error {
	// Send a message de un byte con 0x04 indicating that all bets have been sent
	return p.sendOpcode(OpcodeFinishedBets)
}

func (p *Protocol) requestWinners() error {
	// Send a message indicating that the client wants to consult the winners
	return p.sendOpcode(OpcodeRequestWinners)
}

func (p *Protocol) receiveWinnersList() ([]string, error) {
	// Read number of winners (2 bytes)
	header := make([]byte, 2)
	if _, err := io.ReadFull(p.conn, header); err != nil {
		return nil, err
	}
	numWinners := binary.BigEndian.Uint16(header)

	winners := make([]string, 0, numWinners)
	for i := 0; i < int(numWinners); i++ {
		// Read length of document (1 byte)
		lenBuf := make([]byte, 1)
		if _, err := io.ReadFull(p.conn, lenBuf); err != nil {
			return nil, err
		}
		docLen := int(lenBuf[0])

		// Read document (docLen bytes)
		docBuf := make([]byte, docLen)
		if _, err := io.ReadFull(p.conn, docBuf); err != nil {
			return nil, err
		}
		winners = append(winners, string(docBuf))
	}
	return winners, nil
}

type WinnersResult struct {
	Winners []string
	Waiting bool // true si el servidor responde con OpcodeWait
}

func (p *Protocol) readWinners() (WinnersResult, error) {
	opcode := make([]byte, 1)
	if _, err := io.ReadFull(p.conn, opcode); err != nil {
		return WinnersResult{}, err
	}

	switch opcode[0] {
	case OpcodeWait:
		// El servidor indica que todavia no hay ganadores
		return WinnersResult{Waiting: true}, nil
	case OpcodeWinners:
		// El servidor envía la lista de ganadores
		winners, err := p.receiveWinnersList()
		if err != nil {
			return WinnersResult{}, err
		}
		return WinnersResult{Winners: winners, Waiting: false}, nil
	default:
		return WinnersResult{}, fmt.Errorf("unexpected opcode: %v", opcode[0])
	}
}
